### DISCUSSION TITLE
Cannot open a Twitch with the twitch-disable-ads parameter

### DISCUSSION THREAD

**Discussion by ohmyboroda** [streamlink/streamlink#6027] [Q&A] [answered]
### Description

Hi everyone, I have a issue with opening the twitch with twitch-disable-ads
Could anyone please tell me where I made a mistake, I can't find the answer in the documentation.

```
import cv2
from PIL import Image
from streamlink import Streamlink

# Create Streamlink session
session = Streamlink()

# Setting the twitch-disable-ads parameter
session.set_option("twitch-disable-ads", "true")

# Livestream URL
url = "https://www.twitch.tv/vika_karter"

# Get streams
streams = session.streams(url)
key_list = list(streams.keys())
print(f"Available streams: {key_list}")

# Get the stream in 720p
stream_url = streams["720p60"].url

# Open the stream using OpenCV
cap = cv2.VideoCapture(stream_url)

# Check if the stream was opened successfully
if cap.isOpened():
    # Read a frame from the stream
    success, frame = cap.read()

    # Check if the frame was read successfully
    if success:
        # Convert the OpenCV image to a PIL image
        image = Image.fromarray(cv2.cvtColor(frame, cv2.COLOR_BGR2RGB))

        # Save the image to a file
        image.save("screenshot.png")
        print("Image saved to screenshot.png")
    else:
        print("Failed to read frame from stream")

    # Release the stream
    cap.release()
else:
    print("Failed to open stream")
```

As results
![screenshot](https://github.com/streamlink/streamlink/assets/165786978/cf3aeead-fcc4-44ad-848a-f31d16341931)

**Comment by bastimeyer** [comment, marked as answer]
> `session.set_option("twitch-disable-ads", "true")`

This is not a session option. Valid session options are documented here:
https://streamlink.github.io/api/session.html#streamlink.session.options.StreamlinkOptions

See this answer:
https://github.com/streamlink/streamlink/discussions/5713#discussioncomment-7792946

**Reply by ohmyboroda** [reply]
@bastimeyer sorry for the stupid questions, I've only recently started learning Python, but I don't understand what I'm doing wrong. Looked at your advice in another thread and I can't seem to figure it out.

```
import cv2
from PIL import Image

from streamlink import Streamlink
from streamlink.options import Options

# Create Streamlink session
session = Streamlink()
options = Options()

# Twitch plugin settings
session.plugins.load_builtin()
options.set("twitch-disable-ads", True)
# print(f"Twitch plugins loaded: {session.plugins.get_loaded()}")

# Livestream URL
url = "https://www.twitch.tv/twoangrygamerstv"

# Get streams
stream = session.streams(url, options)
key_list = list(stream.keys())
print(f"Available streams: {key_list}")

# Get the stream
stream_url = stream["best"].url

# Open the stream using OpenCV
cap = cv2.VideoCapture(stream_url)

# Check if the stream was opened successfully
if cap.isOpened():
    # Read a frame from the stream
    success, frame = cap.read()

    # Check if the frame was read successfully
    if success:
        # Convert the OpenCV image to a PIL image
        image = Image.fromarray(cv2.cvtColor(frame, cv2.COLOR_BGR2RGB))

        # Save the image to a file
        image.save("screenshot.jpg")
        print("Image saved to screenshot.png")
    else:
        print("Failed to read frame from stream")

    # Release the stream
    cap.release()
else:
    print("Failed to open stream")
```

**Reply by ohmyboroda** [reply]
Oh gosh, got it figured out :) thanks ❤️

**Reply by ohmyboroda** [reply]
@bastimeyer i guess I got excited before my time, 3 out of 4 don't work. Any idea why the setting is not working?

**Reply by bastimeyer** [reply]
> Any idea why the setting is not working?

Because it's not called `twitch-disable-ads`, as you can see here:
https://github.com/streamlink/streamlink/blob/6.7.4/src/streamlink/plugins/twitch.py#L623-L630

CLI argument names don't represent session- or plugin-option names.

Also, why are you trying to load builtin plugins twice? Initializing a new `Streamlink` session already does that, unless you disabled it via the `plugins_builtin=False` constructor keyword.
https://streamlink.github.io/api/session.html#streamlink.session.Streamlink

And just as explained in the answer of the other discussion linked above, you are not using Streamlink's HLS logic at all. You are simply resolving the stream HLS URL via Streamlink. Whatever you do with it afterwards depends entirely on you and the tools you use, so don't be surprised if you see embedded ads when those tools are unable to filter them out. And don't be confused about what the `disable-ads` plugin option actually does, namely filtering out ads in Streamlink's own HLS implementation, as described very clearly here:
- https://streamlink.github.io/cli.html#cmdoption-twitch-disable-ads
- https://streamlink.github.io/cli/plugins/twitch.html#embedded-ads

**Reply by ohmyboroda** [reply]
Thanks for the feedback

> why are you trying to load builtin plugins twice?

It wasn't obvious to me due to my lack of experience

> Because it's not called twitch-disable-ads, as you can see here

Got it, I couldn't initially find that information in the documentation.

> You are simply resolving the stream HLS URL via Streamlink.

Can you explain on this? I'm already redirecting the stream in which the adverts are expected to be skipped, or is it necessary to make a separate setting?

Even if i simplify the code and try to opening the livestream through VLC as it looks through CLI.
I realise that it may seem obvious or simple to you, but I don't understand it yet and I'm trying to figure it out.

```
import subprocess

from streamlink import Streamlink
from streamlink.options import Options

# Create Streamlink session
session = Streamlink()
options = Options()

session.set_option("hls-segment-stream-data", True)
options.set("disable-ads", True)

# Livestream URL
url = "https://www.twitch.tv/recrent"

# Get streams
stream = session.streams(url, options)
key_list = list(stream.keys())
print(f"Available streams: {key_list}")

# Get the stream
stream_url = stream["best"].url

# Open the stream using VLC
subprocess.call(["vlc", stream_url])
```

**Reply by bastimeyer** [reply]
You are using the **resolved HLS stream URL** and **passing it into a different program** (VLC now, OpenCV prior to that), **which doesn't understand how to filter out advertisement/placeholder segments from the HLS playlist**, because this is very much Twitch-specific.

None of the Streamlink session options or plugin options apply, as you are not using [Streamlink's `HLSStream` implementation](https://streamlink.github.io/api/stream.html) which outputs an **MPEG-TS bitstream** of the concatenated HLS segments. The Twitch plugin subclasses `HLSStream` and has extended logic for filtering out ads (as well as using low-latency streaming), depending on the options set when initializing the plugin. You'll have to use the *file-like object* returned by `stream.open()`. Reading from the stream however should only be done in a secondary thread, as you'd block the main thread otherwise. I don't know any APIs of OpenCV, so this is up to you to figure out.

If this is too much for you to implement, then use Streamlink's CLI and pipe the output into FFmpeg and set the respective FFmpeg options in order to capture screenshots, if this is your goal. Alternatively, invoke Streamlink CLI from your Python module (`streamlink_cli` does not have a public Python API) and write its output into a named pipe which OpenCV might be able to read from, similar to a regular file path, but with live data.

Good luck. This is as much as I'm willing to explain here. The source code is freely available for anyone to study.

**Comment by Sarahsenn340** [comment]
Help me fix unblanned on twich chat
